src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  const newState = { ...state };
+
+  for (const keys in actions) {
+    if (actions[keys].type === 'addProperties') {
+      Object.assign(newState, actions[keys].extraData);
+    }
+
+    if (actions[keys].type === 'removeProperties') {
+      const properties = actions[keys].keysToRemove;
+
+      for (let i = 0; i < properties.length; i++) {
+        delete newState[properties[i]];
+      }
+    }
 
+    if (actions[keys].type === 'clear') {
+      for (const key in newState) {
+        delete newState[key];
+      }
+    }
+
+    const updatedState = { ...newState };
```

Comment:
[DONT REPEAT YOURSELF] - If you perform the same action in all `switch` cases, do it just once afterwards. Consider moving the creation of `updatedState` and pushing it to `stateHistory` outside of the conditional blocks.